src/v/net/batched_output_stream.cc

```diff
@@ -37,24 +37,27 @@ already_closed_error(ss::scattered_message<char>& msg) {
 }
 
 ss::future<bool> batched_output_stream::write(ss::scattered_message<char> msg) {
-    if (unlikely(_closed)) {
+    if (unlikely(_gate.is_closed())) {
         return already_closed_error(msg);
     }
-    return ss::with_semaphore(
-      *_write_sem, 1, [this, v = std::move(msg)]() mutable {
-          if (unlikely(_closed)) {
-              return already_closed_error(v);
-          }
-          const size_t vbytes = v.size();
-          return _out.write(std::move(v)).then([this, vbytes] {
-              _unflushed_bytes += vbytes;
-              if (
-                _write_sem->waiters() == 0 || _unflushed_bytes >= _cache_size) {
-                  return do_flush().then([] { return true; });
+    return ss::with_gate(_gate, [this, msg = std::move(msg)]() mutable {
+        return ss::with_semaphore(
+          *_write_sem, 1, [this, v = std::move(msg)]() mutable {
+              if (unlikely(_gate.is_closed())) {
```

Comment:
This condition can't happen now? We are holding the gate?

Reply:
It can happen, the gate is closed, and we eventually wait for it

Reply:
in other words, gate can be held and closed, only after `gate::close()` returns there are no holders anymore

Reply:
Right yeah, I guess it feels weird to not just send this data off.